arwen/host/execution_gas_test.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	vmOutput, err := host.RunSmartContractCall(input)
+
+	verify := NewVMOutputVerifier(t, vmOutput, err)
+	verify.GasUsed(alphaAddress, alphaCallGas)
+	verify.GasUsed(betaAddress, receiverCallGas)
+	verify.GasUsed(gammaAddress, receiverCallGas)
+	verify.Ok().GasRemaining(expectedGasRemaining)
+}
+
+func createTestParentContract(t testing.TB, host *vmHost, imb *mock.InstanceBuilderMock) {
+	gasUsedByParent := uint64(400)
+	gasProvidedToChild := uint64(300)
+
+	parentInstance := imb.CreateAndStoreInstanceMock(t, host, parentAddress, 0)
+	addDummyMethodsToInstanceMock(parentInstance, gasUsedByParent)
+	addForwarderMethodsToInstanceMock(parentInstance, gasUsedByParent, gasProvidedToChild)
+}
+
+func createTestChildContract(t testing.TB, host *vmHost, imb *mock.InstanceBuilderMock) {
+	gasUsedByChild := uint64(200)
+
+	childInstance := imb.CreateAndStoreInstanceMock(t, host, childAddress, 0)
+	addDummyMethodsToInstanceMock(childInstance, gasUsedByChild)
+}
+
+func createMockBuiltinFunctions(tb testing.TB, host *vmHost, world *worldmock.MockWorld) {
+	err := world.InitBuiltinFunctions(host.GetGasScheduleMap())
+	require.Nil(tb, err)
+
+	world.BuiltinFuncs.Container.Add("builtinClaim", &MockClaimBuiltin{
+		AmountToGive: 42,
+		GasCost:      gasUsedByBuiltinClaim,
+	})
+
+	host.protocolBuiltinFunctions = world.BuiltinFuncs.GetBuiltinFunctionNames()
+}
+
+func addDummyMethodsToInstanceMock(instance *mock.InstanceMock, gasPerCall uint64) {
+	instance.AddMockMethod("wasteGas", func() {
+		host := instance.Host
+		host.Metering().UseGas(gasPerCall)
+	})
+}
+
+func addForwarderMethodsToInstanceMock(instance *mock.InstanceMock, gasPerCall uint64, gasToForward uint64) {
+	input := DefaultTestContractCallInput()
+	input.GasProvided = gasToForward
+
+	t := instance.T
+
+	instance.AddMockMethod("execOnSameCtx", func() {
+		host := instance.Host
+		host.Metering().UseGas(gasPerCall)
+
+		arguments := host.Runtime().Arguments()
+		input.CallerAddr = instance.Address
+		input.RecipientAddr = arguments[0]
+		input.Function = string(arguments[1])
+		numCalls := big.NewInt(0).SetBytes(arguments[2]).Uint64()
+
+		for i := uint64(0); i < numCalls; i++ {
+			_, err := host.ExecuteOnSameContext(input)
+			require.Nil(t, err)
+		}
+	})
+
+	instance.AddMockMethod("execOnDestCtx", func() {
+		host := instance.Host
+		host.Metering().UseGas(gasPerCall)
+
+		argsPerCall := 3
+		arguments := host.Runtime().Arguments()
+		if len(arguments)%argsPerCall != 0 {
+			host.Runtime().SignalUserError("need 3 arguments per individual call")
+			return
+		}
+
+		input.CallerAddr = instance.Address
+
+		for callIndex := 0; callIndex < len(arguments); callIndex += argsPerCall {
+			input.RecipientAddr = arguments[callIndex+0]
+			input.Function = string(arguments[callIndex+1])
+			numCalls := big.NewInt(0).SetBytes(arguments[callIndex+2]).Uint64()
+
+			for i := uint64(0); i < numCalls; i++ {
+				_, _, err := host.ExecuteOnDestContext(input)
+				require.Nil(t, err)
+			}
+		}
+	})
```

Comment:
These 2 have quite some similarities.

I think we can reduce some duplicity here.
Knowing argsPerCall we can make sure that `for callIndex := 0; callIndex < len(arguments); callIndex += argsPerCall ` loops only once if execOnSameCtx

About the error taken from **ExecuteOnSameContext or **ExecuteOnDestContext** we can handle that from function parameter

Reply:
Refactoring these tests is addressed by the branch `gas-async-testing`. But you're right, these can be simplified.